File: android/src/main/java/com/plugin/flutter/zsdk/VirtualDeviceUtils.java
package com.plugin.flutter.zsdk;

import android.text.TextUtils;

import com.zebra.sdk.comm.Connection;
import com.zebra.sdk.comm.ConnectionException;
import com.zebra.sdk.printer.SGD;
import com.zebra.sdk.printer.ZebraPrinter;
import com.zebra.sdk.printer.ZebraPrinterFactory;
import com.zebra.sdk.printer.ZebraPrinterLanguageUnknownException;

/**
 * Created by luis901101 on 11/7/23.
 */
class VirtualDeviceUtils {
   /**
    * This changes the printer Virtual Device only if the current Virtual Device on the printer is different that the one to be changed to and reboots the printer for the changes to take effect.
    * Returns true if the change was necessary and applied, false otherwise.
    */
   public static boolean changeVirtualDevice(Connection connection, String virtualDevice) throws Exception {
      if(TextUtils.isEmpty(virtualDevice)) return false;
      if(connection == null) return false;
      if(!connection.isConnected()) connection.open();

      final String currentVirtualDevice = SGD.GET(SGDParams.KEY_VIRTUAL_DEVICE, connection);
      if(TextUtils.isEmpty(currentVirtualDevice) || !virtualDevice.equals(currentVirtualDevice)) {
         SGD.SET(SGDParams.KEY_VIRTUAL_DEVICE, virtualDevice, connection);
         return PrinterUtils.reboot(connection);
      }
      return false;
   }
}

File: ios/Classes/VirtualDeviceUtils.m
//
//  VirtualDeviceUtils.m
//  zsdk
//
//  Created by Luis on 11/7/23.
//

#import "VirtualDeviceUtils.h"

@implementation VirtualDeviceUtils

+ (bool) changeVirtualDevice:(id<ZebraPrinterConnection,NSObject>)connection virtualDevice:(NSString *)virtualDevice {
    NSError *error = nil;
    NSString *command = @"^XA^JUS^XZ";
    @try {
        if ([ObjectUtils isNull:virtualDevice]) return false;
        if(connection == nil) return false;
        if(![connection isConnected]) [connection open];
        
        const NSString *currentVirtualDevice = [SGD GET:SGDParams.KEY_VIRTUAL_DEVICE withPrinterConnection:connection error:&error];
        
        if (![ObjectUtils isNull:error])
            @throw [NSException exceptionWithName:@"Printer Error" reason:[error description] userInfo:nil];
        if([ObjectUtils isNull:currentVirtualDevice] || ![currentVirtualDevice isEqualToString:virtualDevice]){
            [SGD SET:SGDParams.KEY_VIRTUAL_DEVICE withValue:virtualDevice andWithPrinterConnection:connection error:&error];
            if (error != nil)
                @throw [NSException exceptionWithName:@"Printer Error" reason:[error description] userInfo:nil];
            return [PrinterUtils reboot:connection];
        }
    }
    @catch (NSException *e) {
        @throw e;
    }
    return false;
}

@end
